Interop/Tensorflow/src/main/java/org/tribuo/interop/tensorflow/GradientOptimiser.java

```diff
@@ -0,0 +1,296 @@
+    /**
+     * Gradient descent with Nesterov momentum.
+     * <p>
+     * Parameters are:
+     * <ul>
+     *     <li>learningRate - the learning rate.</li>
+     *     <li>momentum - the momentum scalar.</li>
+     * </ul>
+     */
+    NESTEROV("learningRate","momentum"),
+    /**
+     * The Nadam optimiser.
+     * <p>
+     * Parameters are:
+     * <ul>
+     *     <li>learningRate - the learning rate.</li>
+     *     <li>betaOne - the exponential decay rate for the 1st moment estimates.</li>
+     *     <li>betaTwo - the exponential decay rate for the exponentially weighted infinity norm.</li>
+     *     <li>epsilon - a small constant for numerical stability.</li>
+     * </ul>
+     */
+    NADAM("learningRate","betaOne","betaTwo","epsilon"),
+    /**
+     * The AdaDelta optimiser.
+     * <p>
+     * Parameters are:
+     * <ul>
+     *     <li>learningRate - the overall learning rate.</li>
+     *     <li>decay - the decay factor.</li>
+     *     <li>momentum - the momentum scalar.</li>
+     *     <li>epsilon - for numerical stability.</li>
+     * </ul>
+     * This optimiser is currently uncentered.
+     */
+    RMSPROP("learningRate","decay","momentum","epsilon");
+
+    private final Set<String> args;
+
+    /**
+     * Construct the enum, storing the optimiser argument names.
+     * @param args The optimiser argument names.
+     */
+    private GradientOptimiser(String... args) {
+        this.args = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(args)));
+    }
+
+    /**
+     * An unmodifiable view of the parameter names used by this gradient optimiser.
+     * @return The parameter names.
+     */
+    public Set<String> getParameterNames() {
+        return args;
+    }
+
+    /**
+     * Checks that the parameter names in the supplied set are an exact
+     * match for the parameter names that this gradient optimiser expects.
+     * @param paramNames The gradient optimiser parameter names.
+     * @return True if the two sets intersection and union are equal.
+     */
+    public boolean validateParamNames(Set<String> paramNames) {
+        return (args.size() == paramNames.size()) && args.containsAll(paramNames);
+    }
+
+    /**
+     * Applies the optimiser to the graph and returns the optimiser step operation.
+     * @param graph The graph to optimise.
+     * @param loss The loss to minimise.
+     * @param optimiserParams The optimiser parameters.
+     * @param <T> The loss type (most of the time this will be {@link TFloat32}.
+     * @return The optimiser step operation.
+     */
+    public <T extends TNumber> Op applyOptimiser(Graph graph, Operand<T> loss, Map<String,Float> optimiserParams) {
+        if (!validateParamNames(optimiserParams.keySet())) {
+            throw new IllegalArgumentException("Invalid optimiser parameters, expected " + args.toString() + ", found " + optimiserParams.keySet().toString());
+        }
+        Optimizer optimiser;
```

Comment:
Are we resigned to British/American spelling inconsistency on new code, or do we have a target convention that we'd like to stick to?

Reply:
I think Tribuo is mostly British spellings, so that's what we've got here.